### ISSUE TITLE
bio.h syntax error line 993 - openssl-3.2.0-alpha2 - mingw on Windows - g++.exe (Rev7, Built by MSYS2 project) 13.1.0

### ISSUE THREAD

**Issue by chr0meice2** [openssl/openssl#22296]
Was able to compile static library fine for openssl-3.2.0-alpha2 on mingw/Windows.  Compiling the static lib itself isn't the problem.

Then tried making a basic app after including <openssl/ssl.h> but got this syntax error

**c:\openssl-3.2.0-alpha2\include/openssl/bio.h:993:29: error: expected ')' before ';' token
  993 |                       int (*gets) (BIO *, char *, int));**

I had to comment out the two lines below in bio.h to get my EXE to compile, but no idea why this issue happening in first place

```c

int (*BIO_meth_get_puts(const BIO_METHOD *biom)) (BIO *, const char *);
int BIO_meth_set_puts(BIO_METHOD *biom,
                      int (*puts) (BIO *, const char *));
int (*BIO_meth_get_gets(const BIO_METHOD *biom)) (BIO *, char *, int);
**//int BIO_meth_set_gets(BIO_METHOD *biom,
 //                     int (*gets) (BIO *, char *, int));**
long (*BIO_meth_get_ctrl(const BIO_METHOD *biom)) (BIO *, int, long, void *);
int BIO_meth_set_ctrl(BIO_METHOD *biom,
                      long (*ctrl) (BIO *, int, long, void *));
int (*BIO_meth_get_create(const BIO_METHOD *bion)) (BIO *);
int BIO_meth_set_create(BIO_METHOD *biom, int (*create) (BIO *));

```

My compiler is mingw on Windows

# g++ --version
g++.exe (Rev7, Built by MSYS2 project) 13.1.0

Is this compiler bug? I am compiling with g++  -Wall -Wextra -pedantic  -Os -s

**Comment by mattcaswell** [comment]
Hmmm. I don't see anything obviously wrong with that line. Perhaps there is some kind of symbol clash with some symbol defined elsewhere in a system/other header file. I'm looking at `gets` particularly. What happens if you rename `gets` to some other symbol (it actually could be anything - the name is unimportant in that context), e.g.

````C
int BIO_meth_set_gets(BIO_METHOD *biom,
                      int (*ossl_gets) (BIO *, char *, int));
````

**Comment by tom-cosgrove-arm** [comment]
The other thought is a missing `extern "C"` somewhere - adding that just before `int BIO_meth_set_gets` would see if that is the problem

**Comment by mattcaswell** [comment]
bio.h does already have the `extern "C"` present so this shouldn't be a problem:

https://github.com/openssl/openssl/blob/8f67c6bb7cab70bbcc231ee3e18d140a2857ebdb/include/openssl/bio.h.in#L35-L37

**Comment by nhorman** [comment]
FWIW, I'd identify the full compilation command in the makefile triggering the error, and run it with just the preprocessor (the -E option on gcc, not sure if its the same on mingw), and examine the preprocessor converted output.  could be that windows defines gets as a macro somewhere and its getting token replaced.

**Comment by chr0meice2** [comment]
the issue goes away if changing "gets" to be "gets2".  So yeah it is something with "gets" conflicting... I suggest just change that to be something else.

**Comment by levitte** [comment]
We've seen commentary before (years ago is last I recall) that we really shouldn't use function argument names that are easily conflicting, or at all.  In other words, the advice that I recall is that we should adopt one of these standards for declaring public functions:

- Variant with no argument names at all (I personally dislike it):
  ``` C
  int BIO_meth_set_gets(BIO_METHOD *,
                        int (*) (BIO *, char *, int));
  ```
- Variant with names as comments:
  ``` C
  int BIO_meth_set_gets(BIO_METHOD * /* biom */,
                        int (* /* gets */) (BIO *, char *, int));
  ```
- Variant with names having a project prefix:
  ``` C
  int BIO_meth_set_gets(BIO_METHOD *ossl_biom,
                        int (*ossl_gets) (BIO *, char *, int));
  ```

As this could be done quite broadly, maybe we should discuss this a bit more broadly too...

**Comment by t8m** [comment]
I agree that broader discussion should be done, but it should not prevent us doing a fix right now. (Fix in https://github.com/openssl/openssl/pull/22316)

**Comment by levitte** [comment]
I start a broader discussion

**Comment by slontis** [comment]
Change puts() also if you are going to do this change

**Comment by paulidale** [comment]
puts() is considered _safe_.
gets() is not considered _safe_.

I suspect some magic is being done to prevent use of gets().

**Comment by mspncp** [comment]
@chr0meice2 as stated elsewhere (see https://github.com/openssl/openssl/pull/22316#pullrequestreview-1666526875 and https://github.com/openssl/openssl/pull/22316#issuecomment-1754771618), I would like to see some evidence that the problem is indeed caused by system headers and not by your application.  Could you please


- locate and show us the problematic \#define in the system headers, and
- show us the preprocessor output, [as suggested by @nhorman](https://github.com/openssl/openssl/issues/22296#issuecomment-1751350436) ?
